### PR TITLE
RYA-246-Query-Export-Strategy

### PR DESCRIPTION
**Pull request by meiercaleb**
## Description
Updated Fluo App API so that an ExportStrategy can be specified when a user registers the query with Fluo.  The QueryResultObserver reads the ExportStrategy from the QueryMetadata and the ExportManager then applies the appropriate Exporter(s) based on the QueryType and the ExportStrategy.

### Tests
There were a number of Fluo ITs in place that test this added functionality.

### Links
[Jira](https://issues.apache.org/jira/browse/RYA-246)

### Checklist
- [ ] Code Review
- [ X] Squash Commits

#### People To Reivew
@amihalik @pujav65 @isper3at

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/425/

File: common/rya.api/src/main/java/org/apache/rya/api/client/CreatePCJ.java [hunk 1]
```diff
@@ -27,8 +29,28 @@
...
+     *
+     * @param instanceName - Indicates which Rya instance will create and maintain
+     *   the PCJ. (not null)
+     * @param sparql - The SPARQL query that will be maintained. (not null)
+     * @param strategies - The export strategies used to export results for this query
+     * @return The ID that was assigned to this newly created PCJ.
+     * @throws InstanceDoesNotExistException No instance of Rya exists for the provided name.
+     * @throws RyaClientException Something caused the command to fail.
+     */
+    public String createPCJ(final String instanceName, String sparql, Set<ExportStrategy> strategies) throws InstanceDoesNotExistException, RyaClientException;
```
Comment (amihalik) [inline common/rya.api/src/main/java/org/apache/rya/api/client/CreatePCJ.java]: I can export a result both to Rya and Kafka?

Comment (meiercaleb) [inline common/rya.api/src/main/java/org/apache/rya/api/client/CreatePCJ.java]: Yes.  That was primarily the purpose of this PR.  To allow the user to specify an ExportStrategy on a query by query basis.  Previously, the FluoApp exported using all Exporters that were available, which meant it exported to a RYA PCJ and to Kafka.  There are certain queries where only one type of Exporter is available -- for queries with Aggregations, you can only export to Kafka -- this is a by product of how results are written to the Accumulo table.  This is something that should be fixed going forward.

File: extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/kafka/KafkaBindingSetExporterParameters.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    public boolean getUseKafkaBindingSetExporter() {
+        return getBoolean(params, CONF_USE_KAFKA_BINDING_SET_EXPORTER, false);
+    }
+    
+    /**
+     * 
+     * @param serializer - Used for Serializing BindingSets pushed to Kafka
+     */
+    public void setKafkaBindingSetSerializer(String serializer) {
+        params.put(CONF_KAFKA_BINDING_SET_SERIALIZER, serializer);
```
Comment (meiercaleb) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/kafka/KafkaBindingSetExporterParameters.java]: Null check.

File: extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/kafka/KafkaExportParameterBase.java [hunk 3]
```diff
@@ -31,29 +32,25 @@
...
-     * @param isExportToKafka
-     *            - {@code True} if the Fluo application should export
-     *            to Kafka; otherwise {@code false}.
+     * Sets the bootstrap servers for reading from and writing to Kafka
+     * @param bootstrapServers - connect string for Kafka brokers
      */
-    public void setExportToKafka(final boolean isExportToKafka) {
-        setBoolean(params, CONF_EXPORT_TO_KAFKA, isExportToKafka);
+    public void setKafkaBootStrapServers(String bootstrapServers) {
+        params.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
```
Comment (meiercaleb) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/kafka/KafkaExportParameterBase.java]: Null check.

File: extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/kafka/KafkaExportParameterBase.java [hunk 4]
```diff
@@ -31,29 +32,25 @@
...
+    
     /**
-     * @return {@code True} if the Fluo application should export to Kafka; otherwise
-     *         {@code false}. Defaults to {@code false} if no value is present.
+     * @return Connect string for Kafka servers
      */
-    public boolean isExportToKafka() {
-        return getBoolean(params, CONF_EXPORT_TO_KAFKA, false);
+    public String getKafkaBootStrapServers() {
+        return params.get(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG);
```
Comment (meiercaleb) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/kafka/KafkaExportParameterBase.java]: Docs.  Maybe return optional here.

File: extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/kafka/KafkaSubGraphExporterParameters.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    public boolean getUseKafkaSubgraphExporter() {
+        return getBoolean(params, CONF_USE_KAFKA_SUBGRAPH_EXPORTER, false);
+    }
+    
+    /**
+     * 
+     * @param serializer - Used for Serializing RyaSubGraphs pushed to Kafka
+     */
+    public void setKafkaSubGraphSerializer(String serializer) {
+        params.put(CONF_KAFKA_SUBGRAPH_SERIALIZER, serializer);
```
Comment (meiercaleb) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/kafka/KafkaSubGraphExporterParameters.java]: Null  check.

File: extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/rya/RyaSubGraphExportParameters.java [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    
+    public boolean getUseRyaSubGraphExporter() {
+        return getBoolean(params, CONF_USE_RYA_SUBGRAPH_EXPORTER, false);
+    }
+    
+    /**
+     * @param fluoInstance - the Accumulo instance that Fluo is running on
+     */
+    public void setFluoInstanceName(@Nullable String fluoInstance) {
+        params.put(CONF_FLUO_INSTANCE, fluoInstance);
```
Comment (meiercaleb) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/rya/RyaSubGraphExportParameters.java]: Null check.

Comment (isper3at) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/rya/RyaSubGraphExportParameters.java]: why nullcheck if you tagged it Nullable?

File: extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/kafka/KafkaBindingSetExporterParameters.java (hunk 2 above)
Comment (meiercaleb) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/kafka/KafkaBindingSetExporterParameters.java]: Done.

File: extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/kafka/KafkaExportParameterBase.java (hunk 3 above)
Comment (meiercaleb) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/kafka/KafkaExportParameterBase.java]: Done.

File: extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/kafka/KafkaExportParameterBase.java (hunk 4 above)
Comment (meiercaleb) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/kafka/KafkaExportParameterBase.java]: Done.

File: extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/rya/RyaSubGraphExportParameters.java (hunk 6 above)
Comment (meiercaleb) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/rya/RyaSubGraphExportParameters.java]: Removed Nullable.

File: extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/kafka/KafkaSubGraphExporterParameters.java (hunk 5 above)
Comment (meiercaleb) [inline extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/export/kafka/KafkaSubGraphExporterParameters.java]: Done.

File: extras/rya.indexing.pcj/src/main/java/org/apache/rya/indexing/pcj/storage/accumulo/ShiftVarOrderFactory.java [hunk 7]
```diff
@@ -46,6 +46,10 @@
         final Set<String> bindingNames = new SPARQLParser().parseQuery(sparql, null)
                 .getTupleExpr()
                 .getBindingNames();
+        
+//      //Get Binding names and remove any aggregation variables
+//        final Set<String> bindingNames = new AggregateVariableRemover().getNonAggregationVariables(sparql);
```
Comment (meiercaleb) [inline extras/rya.indexing.pcj/src/main/java/org/apache/rya/indexing/pcj/storage/accumulo/ShiftVarOrderFactory.java]: Done.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/433/

File: common/rya.api/src/main/java/org/apache/rya/api/client/CreatePCJ.java [hunk 8]
```diff
@@ -27,8 +29,28 @@
 @DefaultAnnotation(NonNull.class)
 public interface CreatePCJ {
 
+    public static enum QueryType{Construct, Projection, Periodic};
+    public static enum ExportStrategy{Rya, Kafka, NoOpExport};
```
Comment (ejwhite922) [inline common/rya.api/src/main/java/org/apache/rya/api/client/CreatePCJ.java]: Add javadocs for the enums and their types.  Also change the names for the types to all caps with words separated by _.

File: extras/rya.pcj.fluo/pcj.fluo.api/src/main/java/org/apache/rya/indexing/pcj/fluo/api/CreateFluoPcj.java [hunk 9]
```diff
@@ -128,60 +131,92 @@ public CreateFluoPcj(final int spInsertBatchSize, final int joinBatchSize) {
...
      * @param sparql - sparql query String to be registered with Fluo
+     * @param strategies - ExportStrategies used to specify how final results will be handled
      * @param fluo - A connection to the Fluo application that updates the PCJ index. (not null)
      * @return The metadata that was written to the Fluo application for the PCJ.
      * @throws MalformedQueryException The SPARQL query stored for the {@code pcjId} is malformed.
+     * @throws UnsupportedQueryException 
      * @throws PcjException The PCJ Metadata for {@code pcjId} could not be read from {@code pcjStorage}.
      */
-    public FluoQuery createPcj(String sparql, FluoClient fluo) throws MalformedQueryException {
+    public FluoQuery createPcj(String sparql, FluoClient fluo) throws MalformedQueryException, UnsupportedQueryException {
```
Comment (ejwhite922) [inline extras/rya.pcj.fluo/pcj.fluo.api/src/main/java/org/apache/rya/indexing/pcj/fluo/api/CreateFluoPcj.java]: The @params don't match

File: extras/rya.pcj.fluo/pcj.fluo.api/src/main/java/org/apache/rya/indexing/pcj/fluo/api/CreateFluoPcj.java [hunk 10]
```diff
@@ -128,60 +131,92 @@ public CreateFluoPcj(final int spInsertBatchSize, final int joinBatchSize) {
...
      *
      * @param pcjId - Identifies the PCJ that will be updated by the Fluo app. (not null)
      * @param sparql - sparql query String to be registered with Fluo
+     * @param strategy - ExportStrategies used to specify how final results will be handled
      * @param fluo - A connection to the Fluo application that updates the PCJ index. (not null)
      * @return The metadata that was written to the Fluo application for the PCJ.
+     * @throws UnsupportedQueryException 
      * @throws PcjException The PCJ Metadata for {@code pcjId} could not be read from {@code pcjStorage}.
      */
     public FluoQuery createPcj(
```
Comment (ejwhite922) [inline extras/rya.pcj.fluo/pcj.fluo.api/src/main/java/org/apache/rya/indexing/pcj/fluo/api/CreateFluoPcj.java]: The @params don't match

File: common/rya.api/src/main/java/org/apache/rya/api/client/CreatePCJ.java (hunk 8 above)
Comment (meiercaleb) [inline common/rya.api/src/main/java/org/apache/rya/api/client/CreatePCJ.java]: done.

File: extras/rya.pcj.fluo/pcj.fluo.api/src/main/java/org/apache/rya/indexing/pcj/fluo/api/CreateFluoPcj.java (hunk 9 above)
Comment (meiercaleb) [inline extras/rya.pcj.fluo/pcj.fluo.api/src/main/java/org/apache/rya/indexing/pcj/fluo/api/CreateFluoPcj.java]: done.

File: extras/rya.pcj.fluo/pcj.fluo.api/src/main/java/org/apache/rya/indexing/pcj/fluo/api/CreateFluoPcj.java (hunk 10 above)
Comment (meiercaleb) [inline extras/rya.pcj.fluo/pcj.fluo.api/src/main/java/org/apache/rya/indexing/pcj/fluo/api/CreateFluoPcj.java]: done.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/445/<h2>Failed Tests: <span class='status-failure'>3</span></h2><h3><a name='incubator-rya-master-with-optionals-pull-requests/org.apache.rya:rya.prospector' /><a href='https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/445/org.apache.rya$rya.prospector/testReport'>incubator-rya-master-with-optionals-pull-requests/org.apache.rya:rya.prospector</a>: <span class='status-failure'>3</span></h3><ul><li><a href='https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/445/org.apache.rya$rya.prospector/testReport/org.apache.rya.prospector.mr/ProspectorTest/testCount/'><strong>org.apache.rya.prospector.mr.ProspectorTest.testCount</strong></a></li><li><a href='https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/445/org.apache.rya$rya.prospector/testReport/org.apache.rya.prospector.service/ProspectorServiceEvalStatsDAOTest/testCount/'><strong>org.apache.rya.prospector.service.ProspectorServiceEvalStatsDAOTest.testCount</strong></a></li><li><a href='https://builds.apache.org/job/incubator-rya-master-with-optionals-pull-requests/445/org.apache.rya$rya.prospector/testReport/org.apache.rya.prospector.service/ProspectorServiceEvalStatsDAOTest/testNoAuthsCount/'><strong>org.apache.rya.prospector.service.ProspectorServiceEvalStatsDAOTest.testNoAuthsCount</strong></a></li></ul>

**Pull request closed without merging**